src/interface/swarm_device_context.hpp

```diff
@@ -0,0 +1,112 @@
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef INTERFACE_SWARM_DEVICE_CONTEXT_HPP_
+#define INTERFACE_SWARM_DEVICE_CONTEXT_HPP_
+
+namespace parthenon {
+
+class SwarmDeviceContext {
+ public:
+  KOKKOS_FUNCTION
+  bool IsActive(int n) const { return mask_(n); }
+
+  KOKKOS_FUNCTION
+  bool IsOnCurrentMeshBlock(int n) const { return blockIndex_(n) == this_block_; }
+
+  KOKKOS_FUNCTION
+  void MarkParticleForRemoval(int n) const { marked_for_removal_(n) = true; }
+
+  KOKKOS_FUNCTION
+  bool IsMarkedForRemoval(const int n) const { return marked_for_removal_(n); }
+
+  KOKKOS_INLINE_FUNCTION
+  int GetNeighborBlockIndex(const int &n, const double &x, const double &y,
+                            const double &z, bool &is_on_current_mesh_block) const {
+    int i = static_cast<int>(std::floor((x - x_min_) / ((x_max_ - x_min_) / 2.))) + 1;
+    int j = static_cast<int>(std::floor((y - y_min_) / ((y_max_ - y_min_) / 2.))) + 1;
+    int k = static_cast<int>(std::floor((z - z_min_) / ((z_max_ - z_min_) / 2.))) + 1;
+
+    // Something went wrong
+    if (i < 0 || i > 3 || ((j < 0 || j > 3) && ndim_ > 1) ||
+        ((k < 0 || k > 3) && ndim_ > 2)) {
+      PARTHENON_FAIL("Particle neighbor indices out of bounds");
+    }
+
+    // Ignore k,j indices as necessary based on problem dimension
+    if (ndim_ == 1) {
+      blockIndex_(n) = neighborIndices_(0, 0, i);
+    } else if (ndim_ == 2) {
+      blockIndex_(n) = neighborIndices_(0, j, i);
+    } else {
+      blockIndex_(n) = neighborIndices_(k, j, i);
+    }
+
+    is_on_current_mesh_block = (blockIndex_(n) == this_block_);
+
+    return blockIndex_(n);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  int GetMyRank() const { return my_rank_; }
+
+  KOKKOS_INLINE_FUNCTION
+  void Xtoijk(const Real &x, const Real &y, const Real &z, int &i, int &j, int &k) const {
+    if (ndim_ > 0) {
+      i = static_cast<int>(std::floor((x - x_min_) / dx1_)) + ib_s_;
+      if (ndim_ > 1) {
+        j = static_cast<int>(std::floor((y - y_min_) / dx2_)) + jb_s_;
+        if (ndim_ > 2) {
+          k = static_cast<int>(std::floor((z - z_min_) / dx3_)) + kb_s_;
+        } else {
+          k = kb_s_;
+        }
+      } else {
+        j = jb_s_;
+      }
+    } else {
+      i = ib_s_;
+    }
```

Comment:
Does the logic here work as intended?
For example, `k` is only set if `ndim_ > 2`, but even in a 2d sim, we still carry around the third dimension in terms of indices (but `k=0` in that case).
So I had expected a function similar to
```suggestion
      i = static_cast<int>(std::floor((x - x_min_) / dx1_)) + ib_s_;
      j = (ndim_ > 1) ? static_cast<int>(std::floor((y - y_min_) / dx2_)) + jb_s_ : jb_s_;
      k = (ndim_ > 2) ? static_cast<int>(std::floor((z - z_min_) / dx3_)) + kb_s_ : kb_s_;
```

However, if the function is as intended then some inline comments would be great to highlight that it's okay for some vars to not being set.

Reply:
Yes you are right, in 1D sims `k` was not being set which is bad. I could have added a `k = kb_s_;` below the `j = jb_s_` line but your solution is much cleaner so I've gone ahead with that :)